feat(guides): experimental i18n routing (#5187)
withastro/docs: src/content/docs/en/guides/internationalization.mdx

---
title: Internationalization
description: Learn how to use i18n routing in Astro.
i18nReady: false
---

import FileTree from '~/components/FileTree.astro'
import Since from '~/components/Since.astro'

Astro's internationalization (i18n) features allow you to adapt your project for an international audience.

## i18n Routing (experimental) 

<Since v="3.5.0" />

> The experimental i18n router provides a set functionalities to help building localised websites. 

Reviewer: ```suggestion
The experimental i18n router provides a set of functionalities to help building localised websites. 
```

By default, internationalization routing functionalities are all turned off and the user needs to manually opt in them.

## Glossary

The current page will use some specific terminology related to internationalization:
- Locale: a code that represents a language.
- Localized folder: a folder/directory that is named after a locale.

## Basic Usage

> 1. Enable the routing via configuration

Reviewer: Maybe be a bit more verbose here?

```suggestion
1. Enable the routing via the configuration file
```

> 1. Enable the routing via configuration

Reviewer: Other docs pages seem to have the `.` at the end...

```suggestion
1. Enable the routing via configuration.
```

   ```js title="astro.config.mjs"
   import { defineConfig } from "astro/config"
   export default defineConfig({
     experimental: {
       i18n: {
         defaultLocale: "en",
         locales: ["es", "en", "fr"]
       }
     }
   })
   ```

> 2. Create the folders and pages will contain the translated pages

Reviewer: See above.

```suggestion
2. Create the folders and pages will contain the translated pages.
```

    <FileTree>
    - src
      - pages
        - en
          - index.astro
        - es
          - index.astro
        - fr
          - index.astro
    </FileTree>

    :::note
    The localised folder doesn't need to be at the root of the `pages/` folder.
    :::

> 3. Use the virtual module [`astro:i18n`](#virtual-module-astroi18n) that will help you compute the correct localised URLs:

Reviewer: See above.

```suggestion
3. Use the virtual module [`astro:i18n`](#virtual-module-astroi18n) that will help you compute the correct localised URLs.
```

    ```astro title="src/pages/es/index.astro"
    ---
    import { getRelativeLocaleUrl } from "astro:i18n";
    
    const gettingStartedUrl = getRelativeLocaleUrl("es", "getting-started"); // /es/getting-started
    ---
  
    
    <a href={gettingStartedUrl}>Got to getting started</a>
    ```

## Virtual module `astro:i18n`

> This module provides functions that can help you to create URLs using the locales.

Reviewer: ```suggestion
This module provides functions that can help you create URLs using the locales.
```

All the functions below take into account the following options, so the returned paths will change based on their values:

- [`base`](/en/reference/configuration-reference/#base)
- [`trailingSlash`](/en/reference/configuration-reference/#trailingslash)
- [`build.format`](/en/reference/configuration-reference/#buildformat)
- [`site`](/en/reference/configuration-reference/#site)

> If `i18n.routingStrategy` is equals to `'prefix-other-locales'`, when the functions will process the `i18n.defaultLocale` when passed as input, the returned URLs **won't** contain the `i18n.defaultLocale`.  

Reviewer: What do you mean here as "functions"? This sentence doesn't sound right, but not sure what you mean there to offer a suggestion.

### `getRelativeLocaleUrl()`

`getRelativeLocaleUrl(locale: string, path: string, options?: GetLocaleOptions): string`

Use this function to retrieve a relative path for a locale. If the locale doesn't exist, Astro throws an error. 

```astro title="src/pages/en/index.astro"
---
import {getRelativeLocaleUrl} from "astro:i18n";

getRelativeLocaleUrl("fr", ""); // returns /fr
getRelativeLocaleUrl("fr", "getting-started"); // returns /fr/getting-started
getRelativeLocaleUrl("fr_CA", "getting-started", {
  prependWith: "blog"
}); // returns /blog/fr-ca/getting-started

getRelativeLocaleUrl("fr_CA", "getting-started", {
  prependWith: "blog",
  normalizeLocale: false
}); // returns /blog/fr_CA/getting-started

---
```

### `getAbsoluteLocaleUrl()` 

`getAbsoluteLocaleUrl(locale: string, path: string, options?: GetLocaleOptions): string`

> Use this function to retrieve an absolute path for a locale _if [`site`] has a value_. If [`site`] isn't, the function returns a relative URL. If the locale doesn't exist, Astro throws an error.

Reviewer: ```suggestion
Use this function to retrieve an absolute path for a locale _if [`site`] has a value_. If [`site`] isn't defined, the function returns a relative URL. If the locale doesn't exist, Astro throws an error.
```

```astro title="src/pages/en/index.astro"
---
import {getAbsoluteLocaleUrl} from "astro:i18n";

> // If `site` is set to be `https://example.com`

Reviewer: ```suggestion
// If `site` is set to `https://example.com`
```

getAbsoluteLocaleUrl("fr", ""); // returns https://example.com/fr
getAbsoluteLocaleUrl("fr", "getting-started"); // returns https://example.com/fr/getting-started
getAbsoluteLocaleUrl("fr_CA", "getting-started", {
  prependWith: "blog"
}); // returns https://example.com/blog/fr-ca/getting-started

getAbsoluteLocaleUrl("fr_CA", "getting-started", {
  prependWith: "blog",
  normalizeLocale: false
}); // returns https://example.com/blog/fr_CA/getting-started

---
```

### `getRelativeLocaleUrlList()`

> It works like [`getRelativeLocaleUrl`](#getrelativelocaleurl), so it returns a list of relative paths for all the locales.

Reviewer: ```suggestion
It works like [`getRelativeLocaleUrl`](#getrelativelocaleurl), but it returns a list of relative paths for all the locales.
```

`getRelativeLocaleUrlList(locale: string, options?: GetLocaleOptions): string[]`

### `getAbsoluteLocaleUrlList()`

`getAbsoluteLocaleUrlList(locale: string, options?: GetLocaleOptions): string[]`

> It works like [`getAbsoluteLocaleUrl`](#getabsolutelocaleurl), so it returns a list of relative paths for all the locales.

Reviewer: ```suggestion
It works like [`getAbsoluteLocaleUrl`](#getabsolutelocaleurl), but it returns a list of relative paths for all the locales.
```

## Routing strategy 

> The option `i18n.routingStrategy` allows to change the behaviour of the Astro routes

Reviewer: ```suggestion
The option `i18n.routingStrategy` allows changing the behaviour of the Astro routes.
```

Maybe `router` at the end, and not routes? Might be worth doing a rephrase here...

### `'prefix-other-locales'`

```js title="astro.config.mjs" ins={7}
import { defineConfig } from "astro/config"
export default defineConfig({
  experimental: {
    i18n: {
      defaultLocale: "en",
      locales: ["es", "en", "fr"],
      routingStrategy: "prefix-other-locales"
    }
  }
})
```

This is the **default** value. With this option all the routes *without* locales belong to the default locale (`i18n.defaultLocale`), and the rest of the locales must have their locale code (prefix) in the URL.

- You want to have `example.com/content` for the default locale.

> - You want to have `example.com/es/content` for the other locales.

Reviewer: To harmonise the form with the next example.

```suggestion
- You want to have `example.com/[locale]/content` for the other locales.
```

> - You want to have `example.com/` to redirect (302 status code) to `example.come/<i18n.defaultLocale>`.

Reviewer: ```suggestion

```

Isn't this in perfect contradiction with the first point here? If I'm missing something, there is an extra `to` before `redirect`.

### `'prefix-always'`

```js title="astro.config.mjs" ins={7}
import { defineConfig } from "astro/config"
export default defineConfig({
  experimental: {
    i18n: {
      defaultLocale: "en",
      locales: ["es", "en", "fr"],
      routingStrategy: "prefix-always"
    }
  }
})
```

With this option, all the routes must have their locale code (prefix) in URL.

- You want to have `example.com/[locale]/content` for all routes.
- You want `example.com/content` to return a 404 (not found) status code.

## Browser language detection

> When i18n routing is enabled, two new properties are available via [`Astro` global object](/en/reference/api-reference/#astro-global). These values are **only available in SSR**.

Reviewer: ```suggestion
When i18n routing is enabled, two new properties are available via the [`Astro` global object](/en/reference/api-reference/#astro-global). These values are **only available in SSR**.
```

> The values are calculated by deducing the languages supported by the browser - via the [`Accept-Language`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language) header -  and the locales supported by the website using [`i18n.locales`].

Reviewer: ```suggestion
The values are calculated by deducing the languages supported by the browser - via the [`Accept-Language`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language) header - and the locales supported by the website using [`i18n.locales`].
```

> - `Astro.preferredLocale`: it's the locale requested by the browser and supported by the website. The value is `undefined` if:

Reviewer: I'm assuming this is the first match between the user's list and the website's list.

```suggestion
- `Astro.preferredLocale` is the first locale requested by the browser and supported by the website. The value is `undefined` if:
```

  - The browser/user didn't set any preference.
  - The languages of the browser aren't part of [`i18n.locales`].

> - `Astro.preferredLocaleList`: it's the locales requested by the browser and supported by the website. The value is `[]` if the languages of the browser aren't part of [`i18n.locales`]. The value is [`i18n.locales`] if the browser/user didn't specify any preference.

Reviewer: ```suggestion
- `Astro.preferredLocaleList` is a list of the locales requested by the browser and supported by the website. The value is `[]` if the languages of the browser aren't part of [`i18n.locales`]. The value is [`i18n.locales`] if the browser/user didn't specify any preference.
```

  

## Fallback

> The fallback is a feature that allows to redirect a user from one locale to another. This is useful when pages of a locale miss, but you want to still provide some content to the user.

Reviewer: ```suggestion
The fallback is a feature that allows to redirect a user from one locale to another. This is useful when pages of a locale are missing, but you want to still provide some content to the user.
```

         defaultLocale: "en",
         locales: ["en", "es", "pt-br"]
       }
     }
   })
   ```
  
    You can also specify a [`routingStrategy`](#routing-strategy) to customize the URLs for your default language.

2. Organize your content folders with localized content by language, including your `defaultLocale`. Your folder names must match the items in `locales` exactly. `src/pages/index.astro` will now automatically redirect to the `index.astro` file of your default language.
    <FileTree>
    - src
      - pages
        - en
          - about.astro
          - index.astro
        - es
          - about.astro
          - index.astro
        - pt-br
          - about.astro
          - index.astro
        - index.astro  (redirects to the route created for `/en/index.astro`)
    </FileTree>

    :::note
    The localized folders do not need to be at the root of the `/pages/` folder.
    
    Create individual `/[locale]/` folders anywhere within `src/pages/` and Astro's [file-based routing](/en/core-concepts/routing/) will create your pages at corresponding URL paths. 
    
    ```
    src/pages/blog/[locale]/about.astro -> example.com/blog/es/about/
    ```
    :::

> 3. When writing links to pages on your site, use a localized URL. You can write this URL directly, or compute it using the [`getRelativeLocaleURL()`](#getrelativelocaleurl) helper available from the [`astro:i18n` module](#virtual-module-astroi18n). This will always provide the correct, localized route and can help you correctly use, or check, URLs on your site.

Reviewer 2: > When writing links to pages on your site, use a localized URL. 

@ematipico Just checking on a case like:

- `prefix-other-locales` so that a default locale URL route is `example.com/blog/`
- you want to hardcode (not compute) the href and write it by hand
- you write `href="/blog/"` ???  <-- if this is true, then I should NOT say write a localized URL

Is the above example correct? should I change the wording?

Author: That's correct. The URL `/blog` should be reflected in a folder in `src/pages/blog/index.astro`. With `prefix-other-locales`, it's **assumed** that the content _outside_ of a localized folder belongs to the default locale.

> Is the above example correct? should I change the wording?

I think so, we should mention "localized URL" only for those locales that **aren't** the default locale.

### Configure i18n routing

1. Enable the experimental routing option by adding an `i18n` object to your Astro configuration with a [default location (`defaultLocale`) and a list of all languages to support (`locales`)](#defaultlocale-and-locales):

   ```js title="astro.config.mjs"
   import { defineConfig } from "astro/config"
   export default defineConfig({
     experimental: {
       i18n: {
         defaultLocale: "en",
         locales: ["en", "es", "pt-br"]
       }
     }
   })
   ```
  
2. Choose and configure a [`routingStrategy`](#routingstrategy) based on the desired URL path for your `defaultLocale`:

  - `"prefix-other-locales"` (default): URLs in your default language will **not** have a `/[locale]/` prefix. All other locales will.

  - `"prefix-always"`: All URLs, including your default language, will have a `/[locale]/` prefix.

    ```js title="astro.config.mjs" ins={7}
    import { defineConfig } from "astro/config"
    export default defineConfig({
      experimental: {
        i18n: {
          defaultLocale: "en",
          locales: ["es", "en", "fr"],
          routingStrategy: "prefix-always"
        }
      }
    })
    ```

> 3. Organize your content folders with localized content by language. Your folder names must match the items in `locales` exactly, and your folder organization must match the URL paths chosen for your [`routingStrategy`](#routing-strategy).

Reviewer 2: ```suggestion
3. Organize your content folders with localized content by language. Your folder names must match the items in `locales` exactly, and your folder organization must match the URL paths chosen for your [`routingStrategy`](#routingstrategy).
```

        - es
          - about.astro
          - index.astro
        - pt-br
          - about.astro
          - index.astro
    </FileTree>

    :::note
    The localized folders do not need to be at the root of the `/pages/` folder.
    
    Create individual `/[locale]/` folders anywhere within `src/pages/` and Astro's [file-based routing](/en/core-concepts/routing/) will create your pages at corresponding URL paths. 
    :::
  
4. With i18n routing configured, you can now compute links to pages within your site using the [`getRelativeLocaleURL()`](#getrelativelocaleurl) helper available from the [`astro:i18n` module](#virtual-module-astroi18n). This will always provide the correct, localized route and can help you correctly use, or check, URLs on your site. You can also still write the links manually.

    ```astro title="src/pages/es/index.astro"
    ---
    import { getRelativeLocaleUrl } from 'astro:i18n';

    // defaultLocale is "es"
    const aboutURL=getRelativeLocaleUrl("es", "about");
    ---

    <a href="/get-started/">¡Vamos!</a>
    <a href={getRelativeLocaleUrl('es', 'blog')}>Blog</a>
    <a href={aboutURL}>Acerca</a> 
    ```

### `routingStrategy`

> Astro's built-in file-based routing automatically creates URL routes for you based on your file structure within `src/pages/`. When you configure i18n routing, the `routingStrategy` value now allows you to specify your file structure (and corresponding URL paths generated) in order to use helper functions to generate, use and verify the routes in your project.

Reviewer 3: comma nit
```suggestion
Astro's built-in file-based routing automatically creates URL routes for you based on your file structure within `src/pages/`. When you configure i18n routing, the `routingStrategy` value now allows you to specify your file structure (and corresponding URL paths generated) in order to use helper functions to generate, use, and verify the routes in your project.
```

#### `'prefix-other-locales'`

```js title="astro.config.mjs" ins={7}
import { defineConfig } from "astro/config"
export default defineConfig({
  experimental: {
    i18n: {
      defaultLocale: "en",
      locales: ["es", "en", "fr"],
      routingStrategy: "prefix-other-locales"
    }
  }
})
```

> This is the **default** value. Set this option when URLs in your default language will **not** have a `/[locale]/` prefix, and files in your default language exist at the root of `src/pages/`.

Reviewer 3: comma nit
```suggestion
This is the **default** value. Set this option when URLs in your default language will **not** have a `/[locale]/` prefix and files in your default language exist at the root of `src/pages/`.
```

- `src/pages/blog.astro` will produce the route `example.com/blog/` 
- `src/pages/fr/blog.astro` will produce the route `example.com/fr/blog/` 
- If there is no file at `src/pages/es/blog.astro`, then the route `example.com/es/blog/` will 404 unless you specify a [fallback strategy](#fallback).

#### `'prefix-always'`

```js title="astro.config.mjs" ins={7}
import { defineConfig } from "astro/config"
export default defineConfig({
  experimental: {
    i18n: {
      defaultLocale: "en",
      locales: ["es", "en", "fr"],
      routingStrategy: "prefix-always"
    }
  }
})
```

Set this option when all routes will have their `/locale/` prefix in their URL and when all page content files, including those for your `defaultLocale`, exist in a localized folder:

    <FileTree>
    - src
      - pages
        - index.astro
        - en
          - index.astro
          - about.astro
        - es
          - about.astro
          - index.astro
        - pt-br
          - about.astro
          - index.astro
    </FileTree>

> - URLs without a local prefix, (e.g `example.com/blog/`) will return a 404 (not found) status code.

Reviewer 3: nit: `e.g` -> `e.g.`
```suggestion
- URLs without a local prefix, (e.g. `example.com/blog/`) will return a 404 (not found) status code.
```

### `defaultLocale` and `locales`

> Both a default language ([`defaultLocale`](en/reference/configuration-reference/#experimentali18ndefaultlocale)) and a list of all supported languages ([`locales`](en/reference/configuration-reference/#experimentali18nlocales)) must be specified in your `i18n` routing configuration.

Reviewer 2: ```suggestion
Both a default language ([`defaultLocale`](/en/reference/configuration-reference/#experimentali18ndefaultlocale)) and a list of all supported languages ([`locales`](/en/reference/configuration-reference/#experimentali18nlocales)) must be specified in your `i18n` routing configuration.
```

```js title="astro.config.mjs" ins={6,7-9} 
import { defineConfig } from "astro/config"
export default defineConfig({
  experimental: {
    i18n: {
      defaultLocale: "en",
      locales: ["es", "en", "fr"],
      fallback: {
        fr: "es"  
      }
    }
  }
})
```

Astro will ensure that a page is built in `src/pages/fr` for every page that exists in `src/pages/es/`. If the page does not already exist, then a page with a redirect to the corresponding `es` route will be created.

## Virtual module `astro:i18n`

This module provides functions that can help you create URLs using your project's configured locales.

Creating routes for your project with the i18n router will depend on certain configuration values you have set that affect your page routes. When creating routes with these functions, be sure to take into account your individual settings for:

- [`base`](/en/reference/configuration-reference/#base)
- [`trailingSlash`](/en/reference/configuration-reference/#trailingslash)
- [`build.format`](/en/reference/configuration-reference/#buildformat)
- [`site`](/en/reference/configuration-reference/#site)

> Also note that the returned URLs created by these functions for your `defaultLocale` will reflect your `i18n.routingStrategy` configuration. 

Reviewer 3: comma nit 
```suggestion
Also, note that the returned URLs created by these functions for your `defaultLocale` will reflect your `i18n.routingStrategy` configuration.
```

> URLs created when `prefix-always` configured will include a `/lang/` path in the URL. URLs created with `prefix-other-locales` will not include a language prefix.

Reviewer 3: ```suggestion
URLs created when `prefix-always` is configured will include a `/lang/` path in the URL. URLs created with `prefix-other-locales` will not include a language prefix.
```

---
title: Internationalization
description: Learn how to use i18n routing in Astro.
i18nReady: false
---

import FileTree from '~/components/FileTree.astro'
import Since from '~/components/Since.astro'

Astro's internationalization (i18n) features allow you to adapt your project for an international audience.

> ## i18n Routing (experimental) 

Reviewer 3: Nit: on all the other occurrences of experimental features we added "(Experimental)" capitalized.
```suggestion
## i18n Routing (Experimental)
```

- Localized folder: a folder/directory that is named after a locale.

*/}

<Since v="3.5.0" />

Astro's experimental i18n routing allows you to add your multilingual content with support for configuring a default language and URL path convention, computing relative page URLs, and accepting preferred languages provided by your visitor's browser.

This routing API helps you generate, use, and verify the URLs that your multi-language site produces. Check back and update regularly for the latest changes as this API continues to develop!

### Configure i18n routing

1. Enable the experimental routing option by adding an `i18n` object to your Astro configuration with a [default location (`defaultLocale`) and a list of all languages to support (`locales`)](#defaultlocale-and-locales):

   ```js title="astro.config.mjs"
   import { defineConfig } from "astro/config"
   export default defineConfig({
     experimental: {
       i18n: {
         defaultLocale: "en",
         locales: ["en", "es", "pt_BR"]
       }
     }
   })
   ```
  
    You can also specify a [`routingStrategy`](#routing-strategy) to customize the URLs for your default language.

> 2. Organize your content folders with localized content by language, including your `defaultLocale`. Your folder names must match your lthe items in `locales` exactly. `src/pages/index.astro` will now automatically redirect to the `index.astro` file of your default language.

Reviewer 3: Typo here! Not sure what is the right wording though.

> 2. Organize your content folders with localized content by language, including your `defaultLocale`. Your folder names must match ---> **your lthe** <--- items in `locales` exactly. `src/pages/index.astro` will now automatically redirect to the `index.astro` file of your default language.

Reviewer: Not sure if "your list items" or "the items", probably it came out as a mix ahah

I'd go for the latter.

    <FileTree>
    - src
      - pages
        - en
          - about.astro
          - index.astro
        - es
          - about.astro
          - index.astro
        - pt_BR
          - about.astro
          - index.astro
        - index.astro  (redirects to the route created for `/en/index.astro`)
    </FileTree>

    :::note

>     The localised folders do not need to be at the root of the `/pages/` folder.
>     

Reviewer 3: Typo: `localised` -> `localized`
```suggestion
    The localized folders do not need to be at the root of the `/pages/` folder.
    
```

Reviewer: This is actually valid, are the Docs specifically US English or UK English? I guess other places here use "localize", but my Mac set to UK English autocorrects with an `s`.

Reviewer 2: they are specifically US English!

Each language must be a string (e.g. `"fr"`, `"pt_BR"`), but no particular language format or syntax is enforced while this feature is still experimental and under development. This may be subject to change in future versions.

Your `/[lang]/` folder names must match exactly the `locales` in the list, and every supported language must have its own localized folder. 

Depending on your deploy host, you may discover transformations in URL paths, so check your deployed site to determine the best syntax for your project.

### Routing strategy 

You can choose whether or not pages in your `defaultLocale` should be prefixed with their `/lang/` in their URL path. Every other supported locale will automatically be prefixed.

#### `'prefix-other-locales'`

```js title="astro.config.mjs" ins={7}
import { defineConfig } from "astro/config"
export default defineConfig({
  experimental: {
    i18n: {
      defaultLocale: "en",
      locales: ["es", "en", "fr"],
      routingStrategy: "prefix-other-locales"
    }
  }
})
```

This is the **default** value. With this option, URLs in your default language will **not** have a `/[lang]/` prefix.

> For example, if `i18n.defaultLocal: "en"` is configured: 

Reviewer 3: Typo
```suggestion
For example, if `i18n.defaultLocale: "en"` is configured:
```

  

### Fallback

Astro's i18n routing allows you to configure a **fallback routing strategy**. When a page in one language doesn't exist (e.g. a page that is not yet translated), instead of displaying a 404 page, you can redirect a user from one locale to another on a per-language basis. This is useful when you do not yet have a page for every route, but you want to still provide some content to your visitors.

For example, the configuration below sets `es` as the fallback locale for any missing `fr` routes. This means that a user visiting `example.com/fr/my-page/` will be redirected to and shown the content for `example.com/es/my-page/` instead of being taken to a 404 page when `src/pages/fr/my-page.astro` does not exist.

```js title="astro.config.mjs" ins={6,7-9} 
import { defineConfig } from "astro/config"
export default defineConfig({
  experimental: {
    i18n: {
      defaultLocale: "en",
      locales: ["es", "en", "fr"],
      fallback: {
        fr: "es"  
      }
    }
  }
})
```

> Astro will ensure that a page is built in `src/pages/fr` for every page that exists in `src/pages/es/`. If the page does not already exist, then a page with a redirect to the corresponding `es` route will be be created.

Reviewer 3: ```suggestion
Astro will ensure that a page is built in `src/pages/fr` for every page that exists in `src/pages/es/`. If the page does not already exist, then a page with a redirect to the corresponding `es` route will be created.
```

```

Enabling i18n routing also provides 

## Glossary

The current page will use some specific terminology related to internationalization:
- Locale: a code that represents a language.
- Localized folder: a folder/directory that is named after a locale.

*/}

<Since v="3.5.0" />

Astro's experimental i18n routing allows you to add your multilingual content with support for configuring a default language and URL path convention, computing relative page URLs, and accepting preferred languages provided by your visitor's browser.

This routing API helps you generate, use, and verify the URLs that your multi-language site produces. Check back and update regularly for the latest changes as this API continues to develop!

### Configure i18n routing

1. Enable the experimental routing option by adding an `i18n` object to your Astro configuration with a [default location (`defaultLocale`) and a list of all languages to support (`locales`)](#defaultlocale-and-locales):

   ```js title="astro.config.mjs"
   import { defineConfig } from "astro/config"
   export default defineConfig({
     experimental: {
       i18n: {
         defaultLocale: "en",

>          locales: ["en", "es", "pt_BR"]

Author: I thought again about what we should suggest as locale format, especially for beginners, and I think we should go for `"pt-br"`/`src/pages/pt-br/` instead.

The main reason is that the utilities we provide via `astro:i18n` **normalize** the URLs by default, because that's what most users would want most of the time. This means that if a developer decides to do something like this:

```js
import { getLocaleRelativeUrl } from "astro:i18n";

const aboutUrl = getLocaleRelativeUrl("pt_BR", "about"); // returns /pt-br/about
```

Now `aboutUrl` is `/pt-br/about`, and if a user navigates this URL, they will get a 404 because the file system has the `src/pages/pt_BR/` folder.

Reviewer 2: Sounds good! I will update everything to use xx-yy format!

       }
     }
   })
   ```
  
    You can also specify a [`routingStrategy`](#routing-strategy) to customize the URLs for your default language.

> 2. Organize your content folders with localized content by language, including your `defaultLocale`. Your folder names must match the items in `locales` exactly. `src/pages/index.astro` will now automatically redirect to the `index.astro` file of your default language.

Author: The default `routingStrategy` is `'prefix-other-locales'`, which means that the `defaultLocale` mustn't be part of the URLs.

Hence, we should suggest a different file system structure:

```
- pages
  - index.astro
  - about.astro
  - en/
    - index.astro
    - about.astro
  - pt_BR/
    - index.astro
    - about.astro
```

As a result, we don't have any redirect from `/` to `/en`.

Reviewer 2: Can you clarify, (maybe this changed): I thought that all your FILES MUST ALWAYS be in a localized folder for both options. BUT, the URL routes generated only show the prefix for `prefix-always`.

So, it is also your FOLDER STRUCTURE that must be different for each different option?

Author: > So, it is also your FOLDER STRUCTURE that must be different for each different option?

Yes, that's exactly right. We don't put restrictions on where the localised folders are and their names. The routing strategy **helps** get the URLs in check without doing too many redirects, it's the responsibility of the user to have the physical pages in the correct folders.

This means that with `routingStrategy: prefix-other-locales`, a URL like this `/en/blog` will return a `404`.

I am happy to review the behaviour the strategies as we go. After all, this is experimental. I am sorry for the confusion!

Reviewer 2: Got it! I can update to reflect this!

> 2. Organize your content folders with localized content by language, including your `defaultLocale`. Your folder names must match your lthe items in `locales` exactly. `src/pages/index.astro` will now automatically redirect to the `index.astro` file of your default language.

Reviewer 2: ```suggestion
2. Organize your content folders with localized content by language, including your `defaultLocale`. Your folder names must match the items in `locales` exactly. `src/pages/index.astro` will now automatically redirect to the `index.astro` file of your default language.
```

    <FileTree>
    - src
      - pages
        - en
          - about.astro
          - index.astro
        - es
          - about.astro
          - index.astro
        - pt_BR
          - about.astro
          - index.astro
        - index.astro  (redirects to the route created for `/en/index.astro`)
    </FileTree>

    :::note
    The localized folders do not need to be at the root of the `/pages/` folder.
    

>     Create individual `/[lang]/` folders anywhere within `src/pages/` and Astro's [file-based routing](/en/core-concepts/routing/) will create your pages at corresponding URL paths. 

Author: I think `[lang]` should be changed to `[locale]` instead, because we never mention "lang" before, instead we mention "locale" already, and we use this terminology inside the configuration. Although, **this is my opinion**, you can disregard it  :)

Reviewer 2: yes! we can change them all to `[locale]`! I think that makes perfect sense. Good call.

Your `/[lang]/` folder names must match exactly the `locales` in the list, and every supported language must have its own localized folder. 

Depending on your deploy host, you may discover transformations in URL paths, so check your deployed site to determine the best syntax for your project.

### Routing strategy 

You can choose whether or not pages in your `defaultLocale` should be prefixed with their `/lang/` in their URL path. Every other supported locale will automatically be prefixed.

#### `'prefix-other-locales'`

```js title="astro.config.mjs" ins={7}
import { defineConfig } from "astro/config"
export default defineConfig({
  experimental: {
    i18n: {
      defaultLocale: "en",
      locales: ["es", "en", "fr"],
      routingStrategy: "prefix-other-locales"
    }
  }
})
```

This is the **default** value. With this option, URLs in your default language will **not** have a `/[lang]/` prefix.

For example, if `i18n.defaultLocale: "en"` is configured:

> - `src/pages/en/blog.astro` will produce the route `example.com/blog/` 

Author: This isn't correct. With the `'prefix-other-locales'`, developers must have `src/pages/blog.astro` in their file system, because it's "assumed" that pages outside of a lang/locale folder belong to the default locale. This is related to my comment above.

- `src/pages/fr/blog.astro` will produce the route `example.com/fr/blog/` 
- If there is no file at `src/pages/es/blog.astro`, then the route `example.com/es/blog/` will 404 unless you specify a [fallback strategy](#fallback).

#### `'prefix-always'`

```js title="astro.config.mjs" ins={7}
import { defineConfig } from "astro/config"
export default defineConfig({
  experimental: {
    i18n: {
      defaultLocale: "en",
      locales: ["es", "en", "fr"],
      routingStrategy: "prefix-always"
    }
  }
})
```

With this option, all routes will have their locale code (prefix) in their URL.

> - `src/pages/[lang]/blog.astro` will all reproduce routes of the form `example.com/[locale]/blog/`.

Reviewer: Not sure if this "all" has a reason here... doesn't sound really right to me.

```suggestion
- `src/pages/[lang]/blog.astro` will reproduce routes of the form `example.com/[locale]/blog/`.
```

Reviewer 2: Gotcha. I'll rephrase!

- The URL `example.com/blog/` will return a 404 (not found) status code.

### Browser language detection

Astro's i18n routing combined with one of Astro's  [on-demand server rendering modes](/en/guides/server-side-rendering/) (`output:'server'` or `output:'hybrid'`) allow you to access two properties for browser language detection: `Astro.preferredLocale` and `Astro.preferredLocaleList`.

These combine the browser's `Accept-Langauge` header, and your `locales` to automatically respect your visitor's preferred languages.

- `Astro.preferredLocale`:  Astro can compute a **preferred locale** for your visitor if their browser's preferred locale is included in your `locales` array. This value is undefined if no such match exists.

> - `Astro.preferredLocaleList`: An array of all locales that are both requested by the browser and supported by your website. This produces a list of all compatible languages between your site and your visitor. The value is `[]` if none of the browser's requested languages are found in your `locales` array. If the browser does not specify any preferred languages, then this value will be [`i18n.locales`].

Author: I am not sure if it's worth mentioning it, but I leave the information here, and you decide: the locales coming from `Accept-Language` have their format, e.g. `pt-BR`, although it's possible that developers decide to use a different format inside `i18n.locales`, e.g. `pt_BR`, `pt-br`. 

Astro under to hood does some transformations (no need to mention this) and eventually will return the format **specified in `i18n.locales`**. 

So of `Accept-Language` is `fr-CA;q=0.1, pt-BR;q=0.5", and `i18n.locales` is `["fr_CA", "pt_BR"]`:
- `Astro.preferredLocale` is `"pt_BR"`;
- `Astro.preferredLocaleList` is `["pt_BR", "fr_CA"]`, the order is important because it's kept from `Accept-Language`

Reviewer 2: Let me see what I can do here!
